src/lib/hooks/useUser.tsx

```diff
@@ -1,6 +1,56 @@
-import { useQuery } from "react-query";
 import useAxios from "./useAxios";
 
-export default function useUser() {
-  // TODO: await backend session management
+import { AxiosError } from "axios";
+import Cookies from "js-cookie";
+import { PropsWithChildren, createContext, useContext } from "react";
+import { useQuery, useQueryClient } from "react-query";
+
+export const UserContext = createContext<{
+  user?: User;
+  isLoading: boolean;
+  invalidate: () => void;
+  logout: () => Promise<void>;
+}>({
+  user: undefined,
+  isLoading: true,
+  invalidate: () => {},
+  logout: () => Promise.resolve(),
+});
+
+export const useUser = () => useContext(UserContext);
+
+export function UserProvider({ children }: PropsWithChildren) {
+  const queryClient = useQueryClient();
+  const axios = useAxios();
+  const token = Cookies.get("token");
+  const payload = token ? JSON.parse(atob(token.split(".")[1])) : undefined;
+  const role = payload?.role ?? "employee";
+  const { data: user, isLoading } = useQuery<User | undefined>(
+    [role, "users", "me"],
+    () =>
+      axios
+        .get(`/${role}/users/me`)
+        .then(data => data.data)
+        .catch(err => {
+          if (err instanceof AxiosError) {
+            if (err.response?.status === 403) return;
+          }
+          console.log(err);
+        })
+  );
+
+  const invalidate = () => {
+    queryClient.invalidateQueries([role, "users", "me"]);
```

Comment:
Do not add the role as a key, the user cannot be logged into two roles at the same time so there needs to be a single key referencing a "/users/me" endpoint. You invalidate the entire ['users', 'me] key, regardless of the role. I don't think this affects behavior too much, though.

Reply:
There is something weird happening for logout. No matter what I tried it seems that after the query is invalidated the re-fetching is done with the token set anyways (even though it is explicitly removed in the `logout` function). Do you have any ideas for a fix? I tried creating the `axios` object in the `queryFn` but I think that the whole function might be cached the way it is at the time of creation and until a rerender of `UserProvider` happens (no idea when that is but it takes a few seconds) it doesn't perform the intended query.